File: references/openguardrails-main/openguardrails-main/integrations/agent/opencode/src/wire.ts
/**
 * The OGR v0.8 wire, hand-rolled.
 *
 * There is no SDK layer — the Runtime API is the integration surface
 * (specification/runtime-api.md), and since v0.8 an integration is an API
 * key, nine fields, and ONE decision endpoint. This module is that POST plus
 * the optional heartbeat and the wire types this plugin reads and writes,
 * nothing else: no signing, no batching, no client-side decomposition, and
 * no `/v1/ingest` (removed in v0.8 — evaluate records every event it
 * judges).
 *
 * The base URL is joined with the canonical `/v1/...` paths exactly as the
 * binding requires — a deployment-specific prefix belongs IN the configured
 * base URL (`https://host/api/public/ogr`), never hard-coded here.
 */

/**
 * One v0.8 GuardEvent exactly as the schema requires: every field present,
 * none optional, no extras (`additionalProperties: false`). The four-tuple
 * is required WITH the empty string as the explicit "no assertion" — the
 * integrator answers the identity question; nobody falls into the API-key
 * floor by omission. What v0.6/v0.7 carried and v0.8 removed — versions,
 * coordinates, timestamps, the build id — is derived server-side or lives
 * on the heartbeat, and MUST NOT be sent.
 */
/**
 * WHO REPORTED IT — `"name/version"`, on every event AND on the heartbeat.
 * One constant so the two can never name different builds.
 */
export const INTEGRATION = "ogr-opencode-auto-mode/0.3.0"

export interface WireEvent {
  kind: "step/request" | "step/response"
  /** Fresh per model call, pairing its two halves; the ONE coordinate kept. */
  step_id: string
  agent_id: string
  agent_type: string
  agent_workspace: string
  /**
   * ⚠️ There is no `agent_owner` (removed 2026-08-17). Who is ACCOUNTABLE for an
   * agent is not something a producer can assert — it is a link to a console
   * account an administrator assigns, and a read permission cannot rest on a
   * claim the caller makes about itself.
   */
  agent_user: string
  llm_protocol: "openai.chat" | "openai.responses" | "anthropic.messages" | "canonical"
  payload: Record<string, unknown>
  /**
   * WHO REPORTED IT — `"name/version"`. The one OPTIONAL field on the v0.8
   * wire, stamped by {@link OgrClient.evaluate} so it cannot be forgotten at a
   * construction site.
   *
   * ⚠️ It rode the heartbeat ALONE until 2026-08-17, and that could not answer
   * "which build produced this traffic": a runtime keys its liveness record on
   * the integration NAME (it must, so a rollout updates its row instead of
   * minting a second and reporting the old build as dark), so every replica
   * overwrites the others' version. On the event the build travels with the
   * traffic it produced and nothing can overwrite it.
   */
  integration?: string
}

/** One tool call as a canonical step/response payload carries it. */
export interface WireToolCall {
  id: string
  name: string
  arguments: unknown
}

/** One v0.8 finding — what was found, where, and what it contributed. */
export interface WireFinding {
  category: string
  severity?: "low" | "medium" | "high" | "critical"
  action?: "flag" | "redact" | "block"
  path?: string
  start?: number
  end?: number
  score?: number
  detector?: string
  fp?: string
  whitelisted?: boolean
  subject?: string
}

/** The v0.8 Verdict: two decisions, findings, spans, and the coverage truth. */
export interface WireVerdict {
  event_id: string
  provider: string
  decision: "allow" | "block"
  findings?: WireFinding[]
  modifications?: { spans?: Array<{ path: string; start: number; end: number; replacement: string }> }
  /**
   * Payload paths this verdict could NOT judge. Absent/empty asserts every
   * routed text was judged; under `failMode: "closed"` a non-empty value is
   * "could not look", which is not "found nothing".
   */
  unjudged?: string[]
  latency_ms?: number
}

/** Log sink; the plugin supplies its own so wire noise stays in the host's log. */
export interface WireLog {
  info(message: string): void
  warn(message: string): void
}

/** One resolved runtime connection: where, and as whom. */
export interface RuntimeSource {
  url: string
  apiKey: string
}

/**
 * A fresh `step_id`: a local variable in the loop, never session state.
 * Random per event — this plugin's vantage yields single-event steps, so no
 * id is ever reused across events (the spec's one hard rule for it).
 */
export function mintStepId(): string {
  return globalThis.crypto.randomUUID().replaceAll("-", "")
}

/**
 * The one-POST client (plus the optional heartbeat). The source is a THUNK,
 * re-read on every call, so a connection configured after registration takes
 * effect without a restart.
 */
export class OgrClient {
  /** Heartbeat counters — how an outage-induced observability gap stays visible. */
  eventsSent = 0
  evaluateErrors = 0

  constructor(
    private readonly log: WireLog,
    private readonly source: () => RuntimeSource | null,
    private readonly timeoutMs: () => number,
  ) {}

  /** Whether a runtime is configured RIGHT NOW (the source is live). */
  get enabled(): boolean {
    return this.source() !== null
  }

  private async post(path: string, body: unknown): Promise<Response | null> {
    const src = this.source()
    if (!src) return null
    const base = src.url.endsWith("/") ? src.url.slice(0, -1) : src.url
    const controller = new AbortController()
    const timer = setTimeout(() => controller.abort(), this.timeoutMs())
    try {
      return await fetch(`${base}${path}`, {
        method: "POST",
        headers: {
          "content-type": "application/json",
          authorization: `Bearer ${src.apiKey}`,
        },
        body: JSON.stringify(body),
        signal: controller.signal,
      })
    } finally {
      clearTimeout(timer)
    }
  }

  /**
   * Judge ONE event and return its Verdict, or `null` when no runtime is
   * configured or the call failed (timeout, 429, 5xx, network). Deciding what
   * a missing verdict means is the CALLER's job — that is the deployment's
   * `failMode`, and the degraded-mode spec is explicit that a 429 is an
   * outage, not an allow.
   */
  async evaluate(event: WireEvent): Promise<WireVerdict | null> {
    try {
      // Stamped HERE, not at each construction site: one send path means the
      // build id cannot go missing on one kind of event only.
      const res = await this.post("/v1/evaluate", { ...event, integration: INTEGRATION })
      if (!res) return null
      if (!res.ok) {
        this.evaluateErrors += 1
        this.log.warn(`[openguardrails] evaluate answered ${res.status} — no verdict`)
        return null
      }
      this.eventsSent += 1
      return (await res.json()) as WireVerdict
    } catch (err) {
      this.evaluateErrors += 1
      this.log.warn(`[openguardrails] evaluate failed (${String(err)}) — no verdict`)
      return null
    }
  }

  /**
   * Integration liveness, fire-and-forget: the heartbeat is where the build
   * id lives in v0.8 (it left the event), and its counters are how the
   * runtime tells "agent idle" from "integration went dark". A failed beat
   * is a lost signal, never a lost enforcement.
   */
  async heartbeat(integration: string, agentId: string, intervalS: number): Promise<void> {
    try {
      const res = await this.post("/v1/heartbeat", {
        integration,
        ...agentId ? { agent_id: agentId } : {},
        interval_s: intervalS,
        counters: { events_sent: this.eventsSent, evaluate_errors: this.evaluateErrors },
      })
      if (res && !res.ok) this.log.warn(`[openguardrails] heartbeat answered ${res.status}`)
    } catch (err) {
      this.log.warn(`[openguardrails] heartbeat failed (${String(err)})`)
    }
  }
}

File: references/openguardrails-main/openguardrails-main/integrations/agent/hermes/src/openguardrails_instrumentation_hermes/wire.py
"""The OGR v0.8 wire, hand-rolled.

There is no SDK layer — the Runtime API is the integration surface
(specification/runtime-api.md), and an integration is one decision endpoint:
POST /v1/evaluate while holding an action, plus the optional /v1/heartbeat
for liveness. This module is those two calls over stdlib urllib and the
config they need, nothing else: no enrollment, no signing, no batching
thread, no client-side decomposition (all of it left the protocol with the
SDK layer; /v1/ingest itself was removed in v0.8 — evaluate records).

The base URL is joined with the canonical `/v1/...` paths exactly as the
binding requires — a deployment-specific prefix belongs IN the configured
base URL (`https://host/api/public/ogr`), never hard-coded here.
"""
from __future__ import annotations

import json
import logging
import os
import urllib.error
import urllib.request
from typing import Any

logger = logging.getLogger("ogr-guard.wire")

# Kept in sync with pyproject.toml. Lives on the HEARTBEAT, not on events —
# the build id left the GuardEvent in v0.8; fleet coverage reads it here.
VERSION = "1.0.0"
INTEGRATION = f"openguardrails-instrumentation-hermes/{VERSION}"

# Short and deliberately a hook-path budget, not a model-call budget: every
# evaluate sits between the agent and its next action, so a stalled runtime
# must resolve fast — into the configured fail mode — rather than freeze
# Hermes. Override with OGR_TIMEOUT (seconds).
_DEFAULT_TIMEOUT = 4.0


def _env(name: str, default: str = "") -> str:
    return os.environ.get(name, default).strip()


class OgrClient:
    """The one-endpoint client. Explicit kwargs override the environment:

    OGR_RUNTIME_URL       runtime base URL (unset = no runtime; fail mode decides)
    OGR_API_KEY           organization API key (Authorization: Bearer)
    OGR_AGENT_ID          the four-tuple; all default "" — the explicit
    OGR_AGENT_TYPE          "no assertion", identity then derives from the
    OGR_AGENT_WORKSPACE     API key (the identity floor). agent_type defaults
    OGR_AGENT_USER          to "hermes": the one field this integration DOES
                            know about itself (a label, never an identity).
    OGR_FAIL_MODE         open (default) | closed — what an unanswered
                          evaluate means (specification/degraded-mode.md)
    OGR_TIMEOUT           per-call budget in seconds, default 4.0
    """

    def __init__(
        self,
        *,
        runtime_url: str | None = None,
        api_key: str | None = None,
        agent_id: str | None = None,
        agent_type: str | None = None,
        agent_workspace: str | None = None,
        agent_user: str | None = None,
        fail_mode: str | None = None,
        timeout: float | None = None,
    ) -> None:
        self.runtime_url = (runtime_url if runtime_url is not None
                            else _env("OGR_RUNTIME_URL")).rstrip("/")
        self.api_key = api_key if api_key is not None else _env("OGR_API_KEY")
        # All four always present on every event; "" is a value (the explicit
        # "no assertion"), never an omission — the schema has zero optional fields.
        self.identity: dict[str, str] = {
            "agent_id": agent_id if agent_id is not None else _env("OGR_AGENT_ID"),
            "agent_type": (agent_type if agent_type is not None
                           else _env("OGR_AGENT_TYPE") or "hermes"),
            "agent_workspace": (agent_workspace if agent_workspace is not None
                                else _env("OGR_AGENT_WORKSPACE")),
            "agent_user": agent_user if agent_user is not None else _env("OGR_AGENT_USER"),
        }
        mode = (fail_mode if fail_mode is not None else _env("OGR_FAIL_MODE")).lower() or "open"
        if mode not in ("open", "closed"):
            # A deployment that touched this knob wanted MORE than the default;
            # rounding a typo down to open would silently remove the protection
            # it asked for. The unrecognized value degrades to closed, loudly.
            logger.warning("OGR_FAIL_MODE=%r not recognized — treating as 'closed'", mode)
            mode = "closed"
        self.fail_mode = mode
        try:
            self.timeout = timeout if timeout is not None \
                else float(_env("OGR_TIMEOUT") or _DEFAULT_TIMEOUT)
        except ValueError:
            self.timeout = _DEFAULT_TIMEOUT
        # Heartbeat counters (degraded-mode §2: loud signaling — the
        # evaluate_errors counter is how the runtime learns a PEP went dark).
        self.counters = {"events_sent": 0, "evaluate_errors": 0}

    @property
    def enabled(self) -> bool:
        """Whether a runtime is configured at all. When it is not, every
        evaluate returns None and the FAIL MODE decides — under the default
        (open) the plugin is a pass-through, under closed it denies, which is
        the honest reading of "gate this and the gate is not there"."""
        return bool(self.runtime_url and self.api_key)

    # ------------------------------------------------------------------ #
    # transport
    # ------------------------------------------------------------------ #
    def _post(self, path: str, body: dict[str, Any]) -> dict[str, Any] | None:
        """POST one JSON body; parsed JSON on 2xx, None on ANY failure.
        Never raises: this runs inside Hermes' hook path, and a guard that
        can crash the agent it guards would never be adopted."""
        # default=str: the payload is whatever Hermes handed the hook, and a
        # stray non-JSON fragment must degrade to its string form, not take
        # the whole event (and the hook) down with a TypeError.
        data = json.dumps(body, default=str).encode("utf-8")
        req = urllib.request.Request(
            f"{self.runtime_url}{path}",
            data=data,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {self.api_key}",
            },
            method="POST",
        )
        with urllib.request.urlopen(req, timeout=self.timeout) as resp:
            return json.loads(resp.read().decode("utf-8"))

    # ------------------------------------------------------------------ #
    # the protocol
    # ------------------------------------------------------------------ #
    def evaluate(self, kind: str, step_id: str, llm_protocol: str,
                 payload: dict[str, Any]) -> dict[str, Any] | None:
        """Judge ONE event; the Verdict dict, or None when the runtime could
        not answer (unconfigured, timeout, 401/429/5xx, network, bad JSON).
        Deciding what None means is `blocked()`'s job — that is the
        deployment's fail mode, and the degraded-mode spec is explicit that a
        429 is an outage, not an allow."""
        if not self.enabled:
            return None
        event = {
            "kind": kind,
            "step_id": step_id,
            **self.identity,
            "llm_protocol": llm_protocol,
            "payload": payload,
            # WHO REPORTED IT — the one OPTIONAL v0.8 field. The SAME constant the
            # heartbeat sends: two literals would drift and each would look right.
            #
            # It rode the heartbeat alone until 2026-08-17, which could not answer
            # "which build produced this traffic": a runtime keys its liveness record
            # on the integration NAME (it must, so a rollout updates that row rather
            # than minting a second and reporting the old build as dark), so every
            # replica overwrites the others' version.
            "integration": INTEGRATION,
        }
        try:
            verdict = self._post("/v1/evaluate", event)
        except Exception as exc:  # noqa: BLE001 — every failure is one outcome: no verdict
            self.counters["evaluate_errors"] += 1
            logger.warning("OGR evaluate failed (%s) — no verdict, fail-%s applies",
                           exc, self.fail_mode)
            return None
        self.counters["events_sent"] += 1
        return verdict

    def blocked(self, verdict: dict[str, Any] | None) -> bool:
        """The enforcement question, fail-mode aware.

        Three sizes of "could not look", all governed by the one fail_mode
        (degraded-mode spec): no verdict at all, and a verdict whose
        `unjudged` names paths this event carried. Under the default (open)
        only an explicit block stops the agent; under closed, unprotected
        equals denied.
        """
        if verdict is None:
            return self.fail_mode == "closed"
        if verdict.get("decision") == "block":
            return True
        if self.fail_mode == "closed" and verdict.get("unjudged"):
            # "Could not look" is not "found nothing" — the one assertion a
            # fail-closed deployment rests on is an absent/empty unjudged.
            return True
        return False

    def heartbeat(self) -> bool:
        """Fire-and-forget liveness: lets the runtime tell "agent idle" from
        "integration went dark", and carries the build id + error counters
        (the observability gap an outage causes is visible here, since v0.8
        has no replay channel). Never required, never raises."""
        if not self.enabled:
            return False
        body: dict[str, Any] = {"integration": INTEGRATION, "counters": dict(self.counters)}
        if self.identity["agent_id"]:
            body["agent_id"] = self.identity["agent_id"]
        try:
            self._post("/v1/heartbeat", body)
            return True
        except Exception as exc:  # noqa: BLE001
            logger.warning("OGR heartbeat failed (%s)", exc)
            return False
